File: framework/src/main/java/com/linxiao/framework/common/ErrorMessageUtil.java
package com.linxiao.framework.common;

import com.linxiao.framework.permission.PermissionException;

import java.net.UnknownHostException;
import java.util.HashMap;

import retrofit2.HttpException;

public class ErrorMessageUtil {

    public interface ErrorMessageParser {
        String parseMessage(Throwable e);
    }

    private static final HashMap<Class<? extends Throwable>, ErrorMessageParser> errorMessageMap = new HashMap<>();

    static {
        errorMessageMap.put(UnknownHostException.class, e -> {
            return "unknown host";
        });
        errorMessageMap.put(HttpException.class, e -> {
            return "http error(" + ((HttpException)e).code() + ")";
        });
        errorMessageMap.put(PermissionException.class, e -> {
            return "permission denied";
        });
    }
    public static String getMessageString(Throwable e) {
        ErrorMessageParser parser = errorMessageMap.get(e.getClass());
        if (parser != null) {
            return parser.parseMessage(e);
        }
        return e.getMessage();
    }

    private static void setErrorMessageParser(Class<? extends Throwable> clazz, ErrorMessageParser parser) {
        if (parser == null) {
            return;
        }
        errorMessageMap.put(clazz, parser);
    }
}

File: framework/src/main/java/com/linxiao/framework/common/ErrorMessageUtil.kt
package com.linxiao.framework.common

import com.linxiao.framework.common.ErrorMessageUtil.ErrorMessageParser
import com.linxiao.framework.permission.PermissionException
import retrofit2.HttpException
import java.net.UnknownHostException

object ErrorMessageUtil {

    fun interface ErrorMessageParser<T> {
        fun parseMessage(e: T): String?
    }

    private val errorMessageMap = HashMap<Class<out Throwable>, ErrorMessageParser<Throwable>>()

    init {
        setErrorMessageParser(UnknownHostException::class.java) {
            return@setErrorMessageParser "unknown host"
        }
        setErrorMessageParser(HttpException::class.java) {
            return@setErrorMessageParser "http error(" + (it as HttpException).code() + ")"
        }
        setErrorMessageParser(PermissionException::class.java) {
            return@setErrorMessageParser "permission denied"
        }
    }


    @Suppress("UNCHECKED_CAST")
    fun <T : Throwable> setErrorMessageParser(clazz: Class<T>, parser: ErrorMessageParser<T>?) {
        errorMessageMap[clazz] = ErrorMessageParser {
            return@ErrorMessageParser parser?.parseMessage(it as T)
        }
    }

    @JvmStatic
    fun Throwable.parsedMessage(): String {
        return errorMessageMap[this.javaClass]?.parseMessage(this)
            ?: this.message
            ?: this.javaClass.simpleName
    }
}
